### PR TITLE
[JENKINS-58296] Revert "[JENKINS-53462] Submit form on Firefox as well"

### PR DESCRIPTION
**Pull request by thomasgl-orange**
See [JENKINS-58296](https://issues.jenkins-ci.org/browse/JENKINS-58296) and [Firefox issue 1370630](https://bugzilla.mozilla.org/show_bug.cgi?id=1370630]).

This reverts commit acc4eec688a0594be89f9b5ce860785ed84ac2df.
This change had been made in Jenkins to anticipate Firefox not dispatching
untrusted submit events anymore in v69, but this has actually been backed out.

Since this change seems to have introduced a double-submission bug on forms
with inputs of type "file", reverting sounds like the next logical step.

### Proposed changelog entries

* Revert changes from v2.173/v2.164.3 on forms submission in Jenkins classic UI with Firefox. These were made to anticipate a bugfix in Firefox which has actually been backed out since, and have caused a regression on forms with "file" inputs (double-submission). See [issue 53462](https://issues.jenkins-ci.org/browse/JENKINS-53462), [issue 58296](https://issues.jenkins-ci.org/browse/JENKINS-58296), and [Firefox issue 1370630](https://bugzilla.mozilla.org/show_bug.cgi?id=1370630).

### Submitter checklist

- [X] JIRA issue is well described
- [X] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change).
- [X] Appropriate autotests or explanation to why this change has no tests => _I don't know of any infrastructure for automated testing of Jenkins GUI with real web browsers..._

### Desired reviewers

@daniel-beck (author of reverted commit)

### CODE REVIEW

**Review by res0nance — APPROVED** [review]
LGTM, though I wonder if they will ever reintroduce that patch

**Review by halkeye — APPROVED** [review]
Its certainly right to revert this, but I'm wondering what the longer term fix is.

Does anyone have access to the bug tracker and ask how you can do proper feature detection for this?

**Comment by daniel-beck** [comment]
I would prefer if we found a fix for the specific problem observed around file parameters, rather than doing this.

(My guess is we should be looking at exactly what happens in Jenkins when you have a multipart/form-data form.)

File: war/src/main/webapp/scripts/yui/button/button-debug.js [hunk 1]
```diff
@@ -2719,7 +2719,7 @@ version: 2.9.0
                     method as well.
                 */
               
-                if ((UA.ie || UA.webkit || UA.gecko) && bSubmitForm) {
```
Comment (oleg-nenashev) [inline war/src/main/webapp/scripts/yui/button/button-debug.js]: Add a comment here to reference the historical Gecko issue ?

**Review by oleg-nenashev — APPROVED** [review]
I agree with @daniel-beck that we rather need a proper issue investigation.
Still, fine with me as a hotfix for latest versions. +0

**Review by varyvol — APPROVED** [review]

**Comment by thomasgl-orange** [comment]
> I would prefer if we found a fix for the specific problem observed around file parameters, rather than doing this.
> 
> (My guess is we should be looking at exactly what happens in Jenkins when you have a multipart/form-data form.)

Actually, I did search if anything special was done in Jenkins when submitting in `multipart/form-data` enctype, but without success.
All I have found is the behavior.js trick to set the enctype on the form when there is a file input:
https://github.com/jenkinsci/jenkins/blob/jenkins-2.198/war/src/main/webapp/scripts/hudson-behavior.js#L2592

Some notes on random things I've tried (running master with jenkins-dev:run and -Ddebug.YUI=true to use non-minified YUI files, and testing with FF 70 on Linux):
 - simple job with just one file parameter, selecting "/dev/null" to build the job, it gets POSTed twice.
 - same job when not selecting the input file, gets POSTed only once. The only difference in the multipart request body (apart from the random multiparts separator, of course) is `filename=""` instead of `filename="null"` (in the file0 parameter part).
 - if I change the hudson-behavior.js code above to force of multipart/form-data even when there is no file parameter (adding a `case: "text"` on line 2575), and make a job with only one text parameter, then it still gets POSTed only once too.

So, the form being multipart/form-data is not a sufficient condition to trigger this issue (which doesn't mean it's not part of the equation).

To me, it actually kind of makes sense that the YUI button `submitForm` method submits the form twice after acc4eec688a0594be89f9b5ce860785ed84ac2df (once because of the dispatched event, and once because of the explicit form `submit()` call).
My intuition (and it's really no more than that) is that there is some magic in Firefox which prevents this double-submit to happen in most cases, and that unfortunately the magic doesn't operate when there is a file to read and include in the form data. The best we could do on Jenkins side would thus be to avoid relying on this magic too much. One way of doing it is to avoid calling `submit()` because today it is still redundant, and hope that if Firefox changes its handling of synthetic submit events in the future, we will know in advance in which exact version, and will be able to anticipate it with a change similar to acc4eec688a0594be89f9b5ce860785ed84ac2df but with a condition on `UA.gecko >= X`.

File: war/src/main/webapp/scripts/yui/button/button-debug.js (hunk 1 above)
Comment (thomasgl-orange) [reply war/src/main/webapp/scripts/yui/button/button-debug.js]: OK, I've added 626fa0fff3dfbf2a59b5a0682df76cb7fa99ff66 to document the Firefox (Gecko) situation.

**Comment by oleg-nenashev** [comment]
@daniel-beck are you fine with merging taking the research by @thomasgl-orange and the added comments?

**Comment by daniel-beck** [comment]
I would like some time to investigate this myself. Just reverting doesn't seem like a satisfying solution.

**Review by jvz — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
@daniel-beck what do we for this weekly?

**Comment by krzysztofzawadzki** [comment]
Any decision? @daniel-beck I would suggest to revert and give the time to figure out proper solution without additional complains on bug tracker that issue still exists and FF won't fix the issue in near months.

**Comment by oleg-nenashev** [comment]
I plan to merge it on Thursday if no negative feedback from @daniel-beck or others

**Comment by daniel-beck** [comment]
As per previous agreement with @oleg-nenashev I had until last Friday to come up with an alternative, due to lack of time I have not, so no longer block this.

**Comment by oleg-nenashev** [comment]
OK. We can always revert/change the behavior again, but it makes sense to merge it early to see whetheecit would qualify for LTS backports

**Pull request merged**